pinot-core/src/main/java/org/apache/pinot/core/util/fst/FSTBuilder.java

```diff
@@ -0,0 +1,63 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.util.fst;
+
+import org.apache.lucene.util.IntsRefBuilder;
+import org.apache.lucene.util.fst.Builder;
+import org.apache.lucene.util.fst.FST;
+import org.apache.lucene.util.fst.PositiveIntOutputs;
+import org.apache.lucene.util.fst.Util;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.SortedMap;
+
+public class FSTBuilder {
+    public static final Logger LOGGER =
+            LoggerFactory.getLogger(FSTBuilder.class);
+    private Builder<Long> _builder = new Builder<>(
+            FST.INPUT_TYPE.BYTE4, PositiveIntOutputs.getSingleton());
+    private IntsRefBuilder _scratch = new IntsRefBuilder();
+
+    public void addEntry(String key, Integer value) throws IOException {
+        _builder.add(
+                Util.toUTF16(key, _scratch),
+                value.longValue());
+    }
+
+    public FST done() throws IOException {
+        return _builder.finish();
+    }
+
+    public static FST buildFST(SortedMap<String, Integer> input) throws IOException {
```

Comment:
Explain the algorithm please

Reply:
done, crux of the comments are in RegexpMatcher file